aptly publish crashes on corrupt deb file

I'm using aptly to maintain local mirror(s) of several apt repositories. The workflow is as follows:  update mirror, create timestamped snapshot, publish snapshot.

The `aptly publish snapshot` crashes and leaves ./public directory in inconsistent state when mirror contains corrupt deb file. The public/dists directory is created but empty. The public/pool contain packages aptly processed before crash. Aptly itself does not consider snapshot published.

I have looked into documentation and did not find a way to either automatically verify deb package correctness or filter out single package/version combination from snapshot or publish to work around the issue (external repo and its content is out of my control). 

I can see several ways to solve this issue. All of them assume that we would rather have incomplete, but functional apt repository mirror than no mirror at all:

1) `aptly mirror update` can check for corrupt packages and does not mirror them
2) `aptly snapshot create ...` can check for corrupt packages and not include them
3) `aptly publish ...` can check for corrupt packages and not publish them

Should automated approach be undesirable, manual check of package(s) integrity and purging them from mirror would help here. That way I would be able to update mirror and purge corrupted packages before snapshoting and publishing.

The error I get on corrupted package:

```
You can run 'aptly publish snapshot mesosphere-jessie_2016-09-13T0838' to publish snapshot as Debian repository.
Loading packages...
Generating metadata files and linking package files...
panic: unable to read .tar archive from /srv/aptly/pool/1e/69/marathon_0.15.7-1.0.501.debian81_amd64.deb: unexpected EOF [recovered]
        panic: unable to read .tar archive from /srv/aptly/pool/1e/69/marathon_0.15.7-1.0.501.debian81_amd64.deb: unexpected EOF

goroutine 1 [running]:
panic(0xbc7cc0, 0xc82000e750)
        /usr/lib/go/src/runtime/panic.go:464 +0x3e6
github.com/smira/aptly/cmd.Run.func1(0xc82046dee8)
        /build/aptly-0.9.7/src/github.com/smira/aptly/cmd/run.go:16 +0x7f
panic(0xbc7cc0, 0xc82000e750)
        /usr/lib/go/src/runtime/panic.go:426 +0x4e9
github.com/smira/aptly/deb.(*Package).CalculateContents(0xc820af8f00, 0x7f9415280080, 0xc8200103c0, 0x0, 0x0, 0x0)
        /build/aptly-0.9.7/src/github.com/smira/aptly/deb/package.go:427 +0x1b3
github.com/smira/aptly/deb.(*PackageCollection).loadContents(0xc820010460, 0xc820af8f00, 0x7f9415280080, 0xc8200103c0, 0x0, 0x0, 0x0)
        /build/aptly-0.9.7/src/github.com/smira/aptly/deb/package_collection.go:183 +0x37b
github.com/smira/aptly/deb.(*Package).Contents(0xc820af8f00, 0x7f9415280080, 0xc8200103c0, 0x0, 0x0, 0x0)
        /build/aptly-0.9.7/src/github.com/smira/aptly/deb/package.go:410 +0x75
github.com/smira/aptly/deb.(*ContentsIndex).Push(0xc82043e1e8, 0xc820af8f00, 0x7f9415280080, 0xc8200103c0)
        /build/aptly-0.9.7/src/github.com/smira/aptly/deb/contents.go:26 +0x4a
github.com/smira/aptly/deb.(*PublishedRepo).Publish.func1(0xc820af8f00, 0x0, 0x0)
        /build/aptly-0.9.7/src/github.com/smira/aptly/deb/publish.go:568 +0x8c6
github.com/smira/aptly/deb.(*PackageList).ForEachIndexed(0xc8200139c0, 0xc82046d4d8, 0x0, 0x0)
        /build/aptly-0.9.7/src/github.com/smira/aptly/deb/list.go:169 +0xed
github.com/smira/aptly/deb.(*PublishedRepo).Publish(0xc82019a680, 0x7f9415280080, 0xc8200103c0, 0x7f9415280138, 0xc820135a20, 0xc820206240, 0x7f9415280028, 0xc820202ea0, 0x7f94152800c8, 0xc8200db290, ...)
        /build/aptly-0.9.7/src/github.com/smira/aptly/deb/publish.go:593 +0x113c
github.com/smira/aptly/cmd.aptlyPublishSnapshotOrRepo(0xc8201bcb40, 0xc8201bbce0, 0x1, 0x2, 0x0, 0x0)
        /build/aptly-0.9.7/src/github.com/smira/aptly/cmd/publish_snapshot.go:145 +0x14fe
github.com/smira/commander.(*Command).Dispatch(0xc8201bcb40, 0xc8201bbce0, 0x2, 0x2, 0x0, 0x0)
        /build/aptly-0.9.7/src/github.com/smira/aptly/_vendor/src/github.com/smira/commander/commands.go:305 +0x647
github.com/smira/commander.(*Command).Dispatch(0xc8201bcea0, 0xc8201bbcd0, 0x3, 0x3, 0x0, 0x0)
        /build/aptly-0.9.7/src/github.com/smira/aptly/_vendor/src/github.com/smira/commander/commands.go:283 +0x1f0
github.com/smira/commander.(*Command).Dispatch(0xc8201bd680, 0xc8201bbcc0, 0x4, 0x4, 0x0, 0x0)
        /build/aptly-0.9.7/src/github.com/smira/aptly/_vendor/src/github.com/smira/commander/commands.go:283 +0x1f0
github.com/smira/aptly/cmd.Run(0xc8201bd680, 0xc82009e010, 0x4, 0x4, 0xdbb901, 0x0)
        /build/aptly-0.9.7/src/github.com/smira/aptly/cmd/run.go:40 +0x19a
main.main()
        /build/aptly-0.9.7/src/github.com/smira/aptly/main.go:9 +0x73
```

Same issue seen mirroring the raspian repository (I've seen this reported on Google Groups also).

aptly version: 0.9.7

adminx@server1:~$ aptly publish snapshot rasp_snapshot
Loading packages...
Generating metadata files and linking package files...
panic: unable to ungzip data.tar.gz from /home/adminx/.aptly/pool/b6/f1/linux-source-3.10_3.10.11-1+rpi7_all.deb: gzip: invalid header [recovered]
        panic: unable to ungzip data.tar.gz from /home/adminx/.aptly/pool/b6/f1/linux-source-3.10_3.10.11-1+rpi7_all.deb: gzip: invalid header

goroutine 1 [running]:
panic(0xbc7cc0, 0xc8524cf6a0)
        /usr/lib/go/src/runtime/panic.go:464 +0x3e6
github.com/smira/aptly/cmd.Run.func1(0xc821117f18)
        /build/aptly-0.9.7/src/github.com/smira/aptly/cmd/run.go:16 +0x7f
panic(0xbc7cc0, 0xc8524cf6a0)
        /usr/lib/go/src/runtime/panic.go:426 +0x4e9
github.com/smira/aptly/deb.(*Package).CalculateContents(0xc823183980, 0x7f3c7c688c70, 0xc8201968a0, 0x0, 0x0, 0x0)
        /build/aptly-0.9.7/src/github.com/smira/aptly/deb/package.go:427 +0x1b3
github.com/smira/aptly/deb.(*PackageCollection).loadContents(0xc820196be0, 0xc823183980, 0x7f3c7c688c70, 0xc8201968a0, 0x0, 0x0, 0x0)
        /build/aptly-0.9.7/src/github.com/smira/aptly/deb/package_collection.go:183 +0x37b
github.com/smira/aptly/deb.(*Package).Contents(0xc823183980, 0x7f3c7c688c70, 0xc8201968a0, 0x0, 0x0, 0x0)
        /build/aptly-0.9.7/src/github.com/smira/aptly/deb/package.go:410 +0x75
github.com/smira/aptly/deb.(*ContentsIndex).Push(0xc821c35b30, 0xc823183980, 0x7f3c7c688c70, 0xc8201968a0)
        /build/aptly-0.9.7/src/github.com/smira/aptly/deb/contents.go:26 +0x4a
github.com/smira/aptly/deb.(*PublishedRepo).Publish.func1(0xc823183980, 0x0, 0x0)
        /build/aptly-0.9.7/src/github.com/smira/aptly/deb/publish.go:568 +0x8c6
github.com/smira/aptly/deb.(*PackageList).ForEachIndexed(0xc821b36000, 0xc821117508, 0x0, 0x0)
        /build/aptly-0.9.7/src/github.com/smira/aptly/deb/list.go:169 +0xed
github.com/smira/aptly/deb.(*PublishedRepo).Publish(0xc8201cd380, 0x7f3c7c688c70, 0xc8201968a0, 0x7f3c7c688d28, 0xc8201d4000, 0xc820234340, 0x7f3c7c688c18, 0xc82062b080, 0x7f3c7c688cb8, 0xc8203d6300, ...)
        /build/aptly-0.9.7/src/github.com/smira/aptly/deb/publish.go:593 +0x113c
github.com/smira/aptly/cmd.aptlyPublishSnapshotOrRepo(0xc8201fd9e0, 0xc8201c6860, 0x1, 0x2, 0x0, 0x0)
        /build/aptly-0.9.7/src/github.com/smira/aptly/cmd/publish_snapshot.go:145 +0x14fe
github.com/smira/commander.(*Command).Dispatch(0xc8201fd9e0, 0xc8201c6860, 0x1, 0x2, 0x0, 0x0)
        /build/aptly-0.9.7/src/github.com/smira/aptly/_vendor/src/github.com/smira/commander/commands.go:305 +0x647
github.com/smira/commander.(*Command).Dispatch(0xc8201fdd40, 0xc8201c6850, 0x2, 0x3, 0x0, 0x0)
        /build/aptly-0.9.7/src/github.com/smira/aptly/_vendor/src/github.com/smira/commander/commands.go:283 +0x1f0
github.com/smira/commander.(*Command).Dispatch(0xc8201c85a0, 0xc8201c6840, 0x3, 0x4, 0x0, 0x0)
        /build/aptly-0.9.7/src/github.com/smira/aptly/_vendor/src/github.com/smira/commander/commands.go:283 +0x1f0
github.com/smira/aptly/cmd.Run(0xc8201c85a0, 0xc82000a210, 0x3, 0x3, 0xdbb901, 0x0)
        /build/aptly-0.9.7/src/github.com/smira/aptly/cmd/run.go:40 +0x19a
main.main()
        /build/aptly-0.9.7/src/github.com/smira/aptly/main.go:9 +0x73

--- a/Gomfile
+++ b/Gomfile
@@ -5,6 +5,7 @@ gom 'github.com/cheggaaa/pb', :commit => '2c1b74620cc58a81ac152ee2d322e28c806d81
 gom 'github.com/DisposaBoy/JsonConfigReader', :commit => '33a99fdf1d5ee1f79b5077e9c06f955ad356d5f4'
 gom 'github.com/gin-gonic/gin', :commit => 'b1758d3bfa09e61ddbc1c9a627e936eec6a170de'
 gom 'github.com/go-ini/ini', :commit => 'afbd495e5aaea13597b5e14fe514ddeaa4d76fc3'
+gom 'github.com/h2non/filetype/matchers', :commit => '259d9d2c52bc90dbd7e1999f9da86ee0d104d0ff'
 gom 'github.com/jlaffaye/ftp', :commit => 'fec71e62e457557fbe85cefc847a048d57815d76'
 gom 'github.com/jmespath/go-jmespath', :commit => '0b12d6b521d83fc7f755e7cfc1b1fbdd35a01a74'
 gom 'github.com/julienschmidt/httprouter', :commit => '46807412fe50aaceb73bb57061c2230fd26a1640'

--- a/deb/deb.go
+++ b/deb/deb.go
@@ -2,16 +2,20 @@ package deb
 
 import (
 	"archive/tar"
+	"bufio"
 	"compress/bzip2"
 	"compress/gzip"
 	"fmt"
-	"github.com/mkrautz/goar"
-	"github.com/smira/aptly/utils"
-	"github.com/smira/go-xz"
-	"github.com/smira/lzma"
 	"io"
 	"os"
 	"strings"
+
+	"github.com/h2non/filetype/matchers"
+	"github.com/mkrautz/goar"
+
+	"github.com/smira/aptly/utils"
+	"github.com/smira/go-xz"
+	"github.com/smira/lzma"
 )
 
 // GetControlFileFromDeb reads control file from deb package
@@ -119,29 +123,41 @@ func GetContentsFromDeb(packageFile string) ([]string, error) {
 		}
 
 		if strings.HasPrefix(header.Name, "data.tar") {
+			bufReader := bufio.NewReader(library)
+			signature, err := bufReader.Peek(270)
+
+			var isTar bool
+			if err == nil {
+				isTar = matchers.Tar(signature)
+			}
+
 			var tarInput io.Reader
 
 			switch header.Name {
 			case "data.tar":
-				tarInput = library
+				tarInput = bufReader
 			case "data.tar.gz":
-				ungzip, err := gzip.NewReader(library)
-				if err != nil {
-					return nil, fmt.Errorf("unable to ungzip data.tar.gz from %s: %s", packageFile, err)
+				if isTar {
+					tarInput = bufReader
+				} else {
+					ungzip, err := gzip.NewReader(bufReader)
+					if err != nil {
+						return nil, fmt.Errorf("unable to ungzip data.tar.gz from %s: %s", packageFile, err)
+					}
+					defer ungzip.Close()
+					tarInput = ungzip
 				}
-				defer ungzip.Close()
-				tarInput = ungzip
 			case "data.tar.bz2":
-				tarInput = bzip2.NewReader(library)
+				tarInput = bzip2.NewReader(bufReader)
 			case "data.tar.xz":
-				unxz, err := xz.NewReader(library)
+				unxz, err := xz.NewReader(bufReader)
 				if err != nil {
 					return nil, fmt.Errorf("unable to unxz data.tar.xz from %s: %s", packageFile, err)
 				}
 				defer unxz.Close()
 				tarInput = unxz
 			case "data.tar.lzma":
-				unlzma := lzma.NewReader(library)
+				unlzma := lzma.NewReader(bufReader)
 				defer unlzma.Close()
 				tarInput = unlzma
 			default:

